process/transaction/shardProcess.go

```diff
@@ -119,20 +141,111 @@ func (txProc *txProcessor) ProcessTransaction(tx *transaction.Transaction) error
 	return process.ErrWrongTransaction
 }
 
+func (txProc *txProcessor) createReceiptsWhenFail(tx *transaction.Transaction, acntSnd state.AccountHandler) error {
+	if check.IfNil(acntSnd) {
+		return nil
+	}
+
+	account, ok := acntSnd.(*state.Account)
+	if !ok {
+		return process.ErrWrongTypeAssertion
+	}
+
+	cost := txProc.economicsFee.ComputeFee(tx)
+	if account.Balance.Cmp(cost) < 0 {
+		cost.Set(account.Balance)
+	}
+
+	operation := big.NewInt(0)
+	err := account.SetBalanceWithJournal(operation.Sub(account.Balance, cost))
+	if err != nil {
+		return err
+	}
+
+	err = txProc.increaseNonce(account)
+	if err != nil {
+		return err
+	}
+
+	err = txProc.badTxForwarder.AddIntermediateTransactions([]data.TransactionHandler{tx})
+	if err != nil {
+		return err
+	}
+
+	txHash, err := core.CalculateHash(txProc.marshalizer, txProc.hasher, tx)
+	if err != nil {
+		return err
+	}
+
+	rpt := &receipt.Receipt{
+		Value:   big.NewInt(0).Set(cost),
+		SndAddr: tx.SndAddr,
+		Data:    "invalidTransaction",
+		TxHash:  txHash,
+	}
+
+	err = txProc.receiptForwarder.AddIntermediateTransactions([]data.TransactionHandler{rpt})
+	if err != nil {
+		return err
+	}
+
+	return process.ErrFailedTransaction
+}
+
+func (txProc *txProcessor) createReceiptWithReturnedGas(tx *transaction.Transaction, acntSnd *state.Account) error {
+	if check.IfNil(acntSnd) {
+		return nil
+	}
+	if core.IsSmartContractAddress(tx.RcvAddr) {
+		return nil
+	}
+
+	totalProvided := big.NewInt(0)
+	totalProvided.Mul(big.NewInt(0).SetUint64(tx.GasPrice), big.NewInt(0).SetUint64(tx.GasLimit))
+
+	actualCost := txProc.economicsFee.ComputeFee(tx)
+	refundValue := big.NewInt(0).Sub(totalProvided, actualCost)
+
+	zero := big.NewInt(0)
+	if refundValue.Cmp(zero) == 0 {
+		return nil
+	}
+
+	txHash, err := core.CalculateHash(txProc.marshalizer, txProc.hasher, tx)
+	if err != nil {
+		return err
+	}
+
+	rpt := &receipt.Receipt{
+		Value:   big.NewInt(0).Set(refundValue),
+		SndAddr: tx.SndAddr,
+		Data:    "refundedGas",
+		TxHash:  txHash,
+	}
+
+	err = txProc.receiptForwarder.AddIntermediateTransactions([]data.TransactionHandler{rpt})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
 func (txProc *txProcessor) processTxFee(tx *transaction.Transaction, acntSnd *state.Account) (*big.Int, error) {
 	if acntSnd == nil {
 		return big.NewInt(0), nil
 	}
 
 	err := txProc.economicsFee.CheckValidityTxValues(tx)
 	if err != nil {
+		receiptErr := txProc.createReceiptsWhenFail(tx, acntSnd)
```

Comment:
Why not pass `err` to `txProc.createReceiptsWhenFail()`, so that it is added to the Data field of the Receipt?

Reply:
done.